### PR TITLE
Security Hardening: SEC-1 … SEC-15

### PR DESCRIPTION
**Pull request by mihaivaduva21**
# Security Hardening: SEC-1 … SEC-15

Implements the security findings tracked in `agents/security.json`. Service-to-service
auth is moved from the shared **X-API-KEY** secret to **mTLS with Vault PKI**
(SEC-12), which supersedes the earlier API-key hardening (SEC-6). Adds CI security
scanning (gosec, golangci-lint, govulncheck, Trivy, CodeQL), fixes a set of
correctness/robustness issues in billing flows, and bumps two vulnerable
dependencies.

No database migrations are introduced.

---

## Findings implemented

| ID | Area | Priority | Summary |
|----|------|----------|---------|
| SEC-1 | CCBill IP allowlist | high | Allowlist moved from hard-coded constants to config (`processors.ccbill.allowed_cidrs`), parsed at boot with fail-fast |
| SEC-2 | CI/CD scanning | medium | gosec, golangci-lint, govulncheck, Trivy + CodeQL workflows |
| SEC-3 | Webhook body limits | high | Removed the blanket body-limit exemption for webhooks; added per-processor caps with 413 |
| SEC-4 | Rate limiting | medium | Payload-size (Content-Length) throttling; documented tiers; per-IP/per-user already present |
| SEC-5 | Boot-time config guards | high | Refuse to boot in non-dev when `expected_audience` is unset |
| SEC-6 | API-key timing | high | **Superseded by SEC-12** — the X-API-KEY path was removed, so there is no shared-secret comparison left to harden |
| SEC-7 | Proration rounding | medium | Round **up** to the nearest day; prevents $0 proration near period end |
| SEC-8 | Dunning hard declines | medium | NMI decline-code classification; hard declines cancel immediately, soft declines retry |
| SEC-9 | Admin stale-JWT | medium | DB revocation cross-check on admin **write** endpoints (multi-org), short-TTL cached |
| SEC-10 | NMI upgrade atomicity | high | Inline compensation (refund + cancel) on post-charge failure; idempotent proration charge |
| SEC-11 | SQL hygiene | low | Removed the `db.NewRaw(q.String())` reserialize pattern in audit checks |
| SEC-12 | Service mTLS | medium | Shared X-API-KEY replaced with Vault-PKI **mTLS** (client-cert identity + per-route scopes) |
| SEC-13 | Credit overflow | low | int64 overflow guard on credit deposits |
| SEC-14 | pgx CVEs | high | `jackc/pgx/v5` v5.7.6 → v5.9.2 |
| SEC-15 | x/net CVEs | low | `golang.org/x/net` v0.47.0 → v0.55.0 |

---

### SEC-1 — Configurable CCBill IP allowlist
- `internal/shared/iputil/ccbill.go`: added `Configure([]string) error` (parses all
  CIDRs up front, fails fast on the first invalid entry) guarded by an `RWMutex`;
  `IsValidCCBillIP` reads the configured set. Empty config falls back to the
  documented `DefaultCCBillIPRanges` (fail-closed to published ranges, never
  allow-all).
- `config/config.go`: new `processors.ccbill.allowed_cidrs`; `config.Load` calls
  `iputil.Configure(...)` after validation so an invalid CIDR aborts startup.
- `config.example.yaml`: documents the key and the restart-to-update requirement.

### SEC-2 — CI security scanning
- `.github/workflows/security.yaml`: **gosec** (fail on HIGH+, warn on MEDIUM),
  **golangci-lint** (advisory for now — see `.golangci.yml`), **govulncheck**, and
  **Trivy** (fs scan, HIGH/CRITICAL). Runs on PRs and pushes to `master`.
- `.github/workflows/codeql.yml`: **CodeQL** `security-extended,security-and-quality`
  for Go, hardened (pinned action SHAs, `step-security/harden-runner`), with the
  private-module (`github.com/doujins-org/*`) git config required for autobuild.
- `.golangci.yml`: curated linter set (errcheck, govet, staticcheck, bodyclose,
  noctx, …).

### SEC-3 — Webhook body-size limits
- `internal/http/middleware/security.go`: removed the `/v1/webhooks/*` exemption
  from `BodyLimit`, so the global **1 MiB** cap now backstops webhook routes (this
  is the actual memory-exhaustion fix).
- `internal/http/handlers/webhook.go`: `readLimitedWebhookBody` wraps the body in
  `http.MaxBytesReader` per processor and returns **413** before any processing:
  CCBill 16 KiB, NMI 64 KiB, Stripe 256 KiB. (Sizing note below.)

### SEC-4 — Rate-limiting enhancements
- `internal/http/middleware/ratelimit.go`: **payload-size throttling** — requests
  whose `Content-Length` exceeds a per-bucket ceiling are rejected early with 413
  before any counting. Per-IP and per-authenticated-user limiting already existed
  (both subjects are evaluated; the limiter uses the socket peer IP, not
  `X-Forwarded-For`).
- `docs/rate-limiting.md`: documents subjects, buckets/tiers, payload ceilings, and
  headers.
- *Note:* the `/24` network aggregation from the finding was a “Consider” item and
  is intentionally **not** included.

### SEC-5 — Boot-time config guards
- `config/config.go`: in non-dev environments, `Validate` refuses to boot when
  `expected_audience` is empty (alongside the existing issuers, cors_origins, and
  default-credential checks). The `api_key` guard was dropped — service auth is now
  mTLS (SEC-12), so there is no shared secret to require.

### SEC-6 — Constant-time API-key comparison (superseded)
- Superseded by **SEC-12**. The `X-API-KEY` middleware was removed entirely when
  service auth moved to mTLS, so the constant-time comparison fix no longer applies.
  `internal/http/middleware/apikey.go` no longer exists.

### SEC-12 — Service-to-service mTLS (Vault PKI)
- Service traffic moves to a dedicated **mTLS listener** (`SERVICE_MTLS_PORT=2054`),
  separate from the public port. The shared `X-API-KEY` secret, `PRIVATE_PORT`, and
  `OPENRAILS_API_KEY`/`API_KEY` are removed.
- `internal/http/service_tls.go`: `tls.RequireAndVerifyClientCert` with a configured
  client-CA bundle; the server certificate is reloaded from disk on each handshake so
  Vault/cert-manager renewals roll forward without a restart.
- `internal/http/middleware/mtls.go`: `ServiceMTLSRequired` matches the verified
  client cert's identity (URI SAN → DNS SAN → CN fallback) against the configured
  `clients`, and `RequireServiceScope` enforces per-route scopes (`credits:read`,
  `entitlements:read`, …, with `*` wildcard).
- `config/config.go`: `service_mtls` config + `validateServiceMTLSConfig`
  (fail-fast: port set & distinct from public, cert/key/CA files present, ≥1 client
  identity).
- `docs/mtls-vault-pki.md`: Vault PKI + cert-manager setup, 7-day leaf certs, caller
  config.

### SEC-7 — Proration rounds up
- `internal/modules/checkout/service.go` `CalculateProration`: remaining time is
  computed in seconds and **rounded up** to the next whole day (1 hour left ⇒ 1 day
  charged), preventing effectively-free upgrades near the period boundary.

### SEC-8 — Dunning hard-decline stop
- `internal/modules/subscriptions/dunning.go`: `ClassifyNMIDecline` maps NMI
  `response_code`s to hard vs soft (hard: 201, 204, 220–226, 240, 250–253, 261–263,
  461; everything else soft).
- `internal/integrations/nmi/subscriptions.go`: surfaces `ResponseCode` on
  `ManualRebillResponse`.
- `internal/modules/subscriptions/lifecycle_{types,service}.go`: `FailMembership`
  gains a `HardDecline` flag → immediate cancel, no further retry scheduling.
- `internal/river/jobs_dunning.go`: classifies the decline, logs the code, and
  routes hard declines to immediate cancellation.

### SEC-9 — Admin stale-JWT DB cross-check
- `internal/auth/policy/admin.go` + `admin_cache.go`: for **mutating** admin
  requests in multi-org mode, a short-TTL (30 s) cached DB check blocks callers
  whose account is **positively revoked** (banned/deleted). Read endpoints keep the
  claims-only fast path. Single-org mode was already a live DB role check.
- Fails **open** for users not present in `profiles.users` (external AuthKit
  identities), so legitimate operator admins are never locked out — see Regression
  notes.

### SEC-10 — NMI tier-upgrade atomicity (inline compensation)
- `internal/modules/checkout/service.go` `processUpgrade`:
  - Deterministic proration `OrderID` (`shortHash` of the idempotency key) so a
    retried upgrade reuses the same order reference and NMI dedups the charge.
  - The new local subscription row is created **before** the old subscription is
    cancelled (locally or at NMI), so a failed create leaves the old subscription
    fully intact.
  - `compensateFailedUpgrade` refunds the proration and cancels the new NMI
    subscription on a post-charge DB error.
- **Design choice:** this is inline-compensation only — the optional durable
  marker + background recovery worker were deliberately not included. The only
  uncovered case is a hard process crash in the small window between the NMI charge
  and the DB write, which is logged for alerting.

### SEC-11 — Remove `NewRaw(q.String())` reserialize
- `internal/audit/checks_subscription_state.go`,
  `internal/audit/checks_subscription_entitlement.go`: scan the built query
  directly (`q.Scan(ctx, &results)`), eliminating the bind-dropping reserialize
  pattern. No `fmt.Sprintf`/interpolation exists inside any audit `NewRaw`.

### SEC-13 — Credit deposit overflow guard
- `internal/modules/credits/credits_service.go` `depositTx`: rejects deposits where
  `amount > MaxInt64 - balance` instead of silently wrapping negative.

### SEC-14 / SEC-15 — Dependency bumps
- `jackc/pgx/v5` v5.7.6 → **v5.9.2** (2× HIGH + 2× LOW CVEs).
- `golang.org/x/net` v0.47.0 → **v0.55.0** (5× MEDIUM CVEs).
- `go.mod` / `go.sum` updated; full test suite green.

---

## Regression review & fixes

This PR was reviewed for behavioral regressions. Two were found and fixed within
the PR:

1. **Webhook caps too tight (SEC-3).** Initial caps (CCBill 2 KiB / Stripe & NMI
   5 KiB) would 413 legitimate traffic — Stripe “snapshot” events embed full
   objects (subscriptions, invoices) and routinely exceed 5 KiB, and CCBill
   postbacks carry many fields. Raised to CCBill 16 KiB / NMI 64 KiB / Stripe
   256 KiB. The memory-exhaustion protection is unchanged (global 1 MiB backstop +
   bounded per-processor caps).

2. **Admin check could lock out external admins (SEC-9).** The first version blocked
   any user not found in `profiles.users`, which would 403 legitimate multi-org
   AuthKit operator admins not mirrored locally. Changed to block only on
   **positive** revocation (`IsRevokedUser`: exists AND banned/deleted); unknown
   users fail open.

### Intentional behavioral change (deployment note)
- **SEC-12 (breaking):** service-to-service auth is a **hard cut** from `X-API-KEY`
  to mTLS. `PRIVATE_PORT`, `OPENRAILS_API_KEY`/`API_KEY`, and the `X-API-KEY` header
  no longer grant service-route access. Deployments must enable `service_mtls`
  (cert/key/CA + client identities) and have callers present a Vault-issued client
  cert. See `docs/mtls-vault-pki.md`.

---

## Testing

- `go build ./...`, `go vet ./...`, `gofmt -l` — all clean.
- Full suite: **42/42 packages pass**, zero failures. `-race` clean on the
  concurrency-sensitive packages (`iputil`, `ratelimit`, `auth/policy`).
- New/updated tests: CCBill config allowlist, webhook 413 caps, proration round-up
  edge cases, NMI decline classification, admin verify-cache TTL + mutating-method
  gate, credit-deposit overflow (DB-gated integration), upgrade compensation +
  `shortHash`, body-limit-applies-to-webhooks.

---

## Out of scope
- SEC-4 `/24` network aggregation (a “Consider” item).
- SEC-10 durable marker + background recovery worker (inline compensation chosen
  instead).

### CODE REVIEW

File: internal/http/middleware/apikey.go [hunk 1]
```diff
@@ -39,13 +41,12 @@
...
-	if len(a) != len(b) {
-		return false
-	}
-	var result byte
-	for i := 0; i < len(a); i++ {
-		result |= a[i] ^ b[i]
-	}
-	return result == 0
+	aHash := sha256.Sum256([]byte(a))
+	bHash := sha256.Sum256([]byte(b))
```
[the remark this answers is not in the document]
Comment (mihaivaduva21) [reply internal/http/middleware/apikey.go]: This is a FP, SHA256 is used for hashing in constant time, not for password hash storage.

**Comment by mihaivaduva21** [comment]
@copilot apply review comments from this thread https://github.com/open-rails/openrails/pull/59#pullrequestreview-4405200621
